src/bsoncxx/builder/list/document.hpp

```diff
@@ -0,0 +1,178 @@
+#include <bsoncxx/types/bson_value/value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace builder {
+namespace list {
+using namespace bsoncxx::types;
+
+///
+/// A JSON-like builder for creating BSON documents.
+///
+class document {
+   public:
+    ///
+    /// Default constructor. Creates an empty document.
+    ///
+    document() = default;
+
+    ///
+    /// Creates a document from a key-value pair. Initializes the key with 'key' and the value with
+    /// 'value'. The value may be another document.
+    ///
+    /// @param key
+    ///     the value to initialize the key in the key-value pair.
+    /// @param value
+    ///     the value to initialize the value in the key-value pair.
+    ///
+    /// @note a bsoncxx::types::bson_value::value is direct-initialized from 'value'.
+    ///
+    /// @see bsoncxx::types::bson_value::value.
+    ///
+    template <typename T>
+    document(stdx::string_view key, T&& value) {
+        _core.key_view(key);
+        _core.append(bson_value::value{value});
+    }
+
+    ///
+    /// Creates a document from a key-value pair. Initializes the key with 'key' and the value with
+    /// 'value'. The value may be another document.
+    ///
+    /// @param key
+    ///     the value to initialize the key in the key-value pair.
+    /// @param value
+    ///     the value to initialize the value in the key-value pair.
+    ///
+    /// @note a bsoncxx::types::bson_value::value is direct-initialized from 'value'.
+    ///
+    /// @see bsoncxx::types::bson_value::value.
+    ///
+    template <typename T>
+    document(stdx::string_view key, const T& value) {
+        _core.key_view(key);
+        _core.append(bson_value::value{value});
+    }
+
+    ///
+    /// Constructs the document with the contents of the initializer list 'init'. All documents in
+    /// 'init' are concatenated.
+    ///
+    /// @param init initializer list to initialize the key-value pairs of the document with
+    ///
+    document(std::initializer_list<document> init) {
+        for (auto doc : init) {
+            bson_value::value val{doc};
+            _core.concatenate(val.view().get_document());
+        }
+    }
+
+    ///
+    /// Move constructor. Constructs the document with the contents of other using move semantics.
+    ///
+    /// @param other another document to use as source to initialize the document with
+    ///
+    document(document&& other) noexcept = default;
+
+    ///
+    /// Replaces the contents with those of other using move semantics.
+    ///
+    /// @param other another document to use as source to initialize the document with
+    ///
+    document& operator=(document&& other) noexcept = default;
+
+    ///
+    /// Copy constructor. Constructs the document with a copy of the contents of other.
+    ///
+    /// @param other another document to use as source to initialize the document with
+    ///
+    document(const document& other) {
```

Comment:
Same question as with the array builder about usefulness of a copy constructor and assignment operator, and when somebody might want to copy a builder

Reply:
Removed.